### PR TITLE
Modification of colours rendering + 256 colours support

### PR DESCRIPTION
**Pull request by Lynesth**
Disclaimer: It isn't finished yet

So this adds support for 256 colors if the terminal supports it (it needs https://github.com/php-school/terminal/pull/8 to work).

For now it only works if such colours are set using their number (0 to 255, see image attached) and doesn't yet fallback to 8 colours if not supported by terminal. I also want to add a full table of names->code association so that we can set them using their name.

It also modifies the way colours escaped sequences are generated so that it only happens as little as possible (when colours are set) and not each time.

It also makes use of the inverted escape sequence to set the colours for selected elements and the reset escape sequence to reset all colours/blod/whatever to their default value.

Let me know what you think so that I continue on this, or not. As with my previous PRs, I'll fix the tests if it's accepted.

List of colours:
ktsqa

Ain't it sexy ?
mintty_2018-05-06_01-17-49

### CODE REVIEW

**Comment by AydinHassan** [comment]
Just skimmed over it and the ideas all seem sounds to me, please continue! pinging @mikeymike here to checkout also, he wrote most of this stuff!

**Comment by Lynesth** [comment]
Ok so I spent some time wondering what would be the best way to fallback to 8 colors scheme from a 256 colors code if the terminal did not support it.

Properly (not using RGB) computing the difference between the colors to automatically select the closest one from the default 8 seems to be a bit expensive, so I went with the $fallback argument. It feel that it also gives more control over it.

On another note, do we need some new Exception class to throw those "Invalid colour code" exceptions ?

**Comment by AydinHassan** [comment]
@lachlankrautz could we just maybe calculate the fallbacks ourself and just store it as code? Like perform the calculations to figure which 8bit colour each of the 256 should map to and store it as a hardcoded array somewhere? that way we don't need to perform expensive computations at runtime

**Comment by Lynesth** [comment]
@AydinHassan Yeah I guess I can do that.
Many colours will end up as the same one (of course going from 256 to 8) so do we still want to be able to set the fallback manually ?

**Comment by Lynesth** [comment]
Closing/Reopening to trigger a Travis rebuild since https://github.com/php-school/terminal/pull/9 was merged.

**Comment by Lynesth** [comment]
There you go: https://gist.github.com/Lynesth/5b8d885a2f3c0bef180c6733813ef156

Where do you want it included ? We would need it both in MenuStyle and in CliMenuBuilder.
Do we make a new class that just has one static function to return the fallback value ?

Let me know.

**Comment by AydinHassan** [comment]
@Lynesth this is really cool! Yeah I agree a class with one static method should work. Maybe be dump it in `PhpSchool\CliMenu\Style`. Name is hard, hmm. Maybe:

```php
class ColourMap {
    public static function map256To8()
    {}
}
```

? 

I'm open for other suggestions though.

We can still fallback manually also, but I guess it will be optional then.

File: src/CliMenuBuilder.php [hunk 1]
```diff
@@ -198,18 +198,36 @@ public function setExitButtonText(string $exitButtonText) : self
         return $this;
     }
 
-    public function setBackgroundColour(string $colour) : self
-    {
-        Assertion::inArray($colour, MenuStyle::getAvailableColours());
+    public function setBackgroundColour($colour, string $fallback = null) : self
+    {
+        if (is_int($colour)) {
+            if ($this->terminal->getColourSupport() < 256) {
```
Comment (AydinHassan) [inline src/CliMenuBuilder.php]: Looks like we can extract this validation as it's the same in `setForegroundColour`. Lets introduce a `validateColour` method.

**Comment by AydinHassan** [comment]
@mikeymike can you take a look at this at some point since you wrote most of this code?

**Pull request merged**